src/rnnlm/arpa-sampling.h

```diff
@@ -0,0 +1,131 @@
+#include <vector>
+
+#include <math.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include "fst/fstlib.h"
+#include "util/common-utils.h"
+#include "lm/arpa-file-parser.h"
+
+namespace kaldi {
+
+class ArpaSampling : public ArpaFileParser {
+ public:
+  friend class ArpaSamplingTest;
+
+  // HistType represents a history
+  typedef std::vector<int32> HistType;
+  // WordToProbsMap represents the words and their probabilities given
+  // an arbitrary history
+  typedef std::unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
+  // NgramType represents the map between a history and the map of existing words
+  // and their probabilities given this history
+  typedef std::unordered_map<HistType, WordToProbsMap, VectorHasher<int32> > NgramType;
+  // HistWeightsType represents the map between histories (with weights) and
+  // their computed weights
+  typedef std::unordered_map<HistType, BaseFloat, VectorHasher<int32> > HistWeightsType;
+
+  // constructor
+  ArpaSampling(ArpaParseOptions options, fst::SymbolTable* symbols)
+     : ArpaFileParser(options, symbols) { 
+       ngram_order_ = 0;
+       num_words_ = 0;
+       bos_symbol_ = "<s>";
+       eos_symbol_ = "</s>";
+       unk_symbol_ = "<unk>";
+  }
+  
+  // This function computes the unigram distribution of all vocab words which are
+  // represented by integers from 1 to num_words_ (0 is reserved for epsilon)
+  // Probabilities of integers not represented by any valid word is 0, e.g.
+  // 0 has probability 0.0
+  void GetUnigramDistribution(std::vector<BaseFloat> *unigram_probs);
+
+  // This function reads in a list of histories with input weights and returns
+  // 1) the non_unigram_probs (maps the higher-than-unigram words to their
+  // corresponding probabilities given the list of histories) and 
+  // 2) a scalar unigram_weight = sum of history_weight * backoff_weight of that history
+  BaseFloat GetDistribution(const std::vector<std::pair<HistType, BaseFloat> > 
+      &histories, std::unordered_map<int32, BaseFloat> *non_unigram_probs);
+ 
+ protected:
+  // ArpaFileParser overrides.
+  virtual void HeaderAvailable(); 
+  virtual void ConsumeNGram(const NGram& ngram);
+  virtual void ReadComplete() {}
+
+ private:
+  // This function returns the log probability of a LM state, [history word],
+  // from the read-in ARPA file if it exists. If the LM state does not exist,
+  // it backs off to a lower order until the updated LM state found.
+  // Note: a LM state is a ngram term in a ARPA language model
+  BaseFloat GetProb(int32 order, int32 word, const HistType& history);
+
+  // This function returns the back-off log probability of a LM state
+  // from the read-in ARPA file 
+  BaseFloat GetBackoffWeight(int32 order, int32 word, const HistType& history);
+
+  // This function returns the computed weights of histories with input weights
+  void ComputeHistoriesWeights(const std::vector<std::pair<HistType, BaseFloat> > 
```

Comment:
Both types are ok for my code if there are no repeated history in the input history list. I did this conversion because I assumed the history list may have repeated history (then loop over those repeated histories are not efficient). If there are no repeated histories, I can remove this function and we don't need to change the interface of GetDistribution().